### PR TITLE
Space - Becca Roach

### PR DESCRIPTION
**Pull request by RebeccaRoach**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | Deciding which classes and which methods "belonged" to those classes was a challenge with this project. Particularly as it pertained to the DateRange class, I was at first unsure if it would be necessary or even very useful to have it as its own class. Though I ended up utilizing DateRange in Reservation and Controller, I can see how that design decision created dependencies in the system. `Refactors.txt` contains notes about how I might get around the need for this class altogether. |
What was a design decision you made that changed over time over the project? | One design decision that changed over time was the idea about which class(es) should know about Reservations. At first I thought it would be convenient to have the Controller have a list of all its reservations, as well as individual rooms. However, further thinking on the issue of duplication and potential mishaps in syncing/ maintaining both places of the data led me to rethink this structure. If modifications or cancellations were a required functionality of our Hotel system, that would present an even stronger reason to isolate the data to one place only. Though keeping reservations in single rooms made looping through all reservations (i.e. to tell if the Hotel had availability for a given date range) more difficult, it made Wave 2 requirements (checking for room availability) much more straightforward.
What was a concept you gained clarity on, or a learning that you'd like to share? | I learned more about the power of attr_accessors through this project, both good and bad. On the one hand, they're useful for classes to be able to interact with each other, but on the other hand, having too many accessors make the system more vulnerable to bugs or for users to potentially misuse/ break the code.|
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | A _nominal test_ in this project was for find_available_rooms in the Controller class, "returns only the correct room ids of the Rooms that have availability". For this test, I created a room reservation for Rooms with room_ids evenly divisible by 3. I would expect that room 3 would then be unavailable, but that rooms 1, 2, 4, and 5 would still be available, for example. As this was the expected behavior, this can be considered a nominal case for testing. |
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | An example of an _edge case test_ was a test in Reservation's constructor, it "throws an argument error if given an invalid room_ID value". In this test, I deliberately entered negative values, 0, values > 20, nil, and even a string because, for one reason or another, these were unexpected inputs that would need to raise an ArgumentError. Because such values would be considered invalid for use in this specific hotel program, this represents an _edge case test_. |
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | Though it was challenging, I felt really proud of myself for undertaking writing pseudocode and being committed to TDD all throughout this project! It took me a significantly longer amount of time to write all my tests before writing methods, but it was a worthwhile process that helped me catch sneaky bugs and guided me overall to practice better development habits than I have ever had before. :) I did not get as far in the waves as I had hoped, but the experience I got with the TDD process and with the concepts of Object Oriented design was significant and worth every bit of effort. |

### CODE REVIEW

File: lib/reservation.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+      @date_range = date_range
+      @room_id = room_id
+
+      raise ArgumentError if room_id < 1 || room_id > 20
+    end
+
+    def cost
+      num_nights = (@date_range.end_date - @date_range.start_date).to_i
+      fixed = 200
```
Comment (dHelmgren) [inline lib/reservation.rb:18]: Does this need to be a variable, or could it be a constant?

File: lib/hotel_controller.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+      return available
+    end
+
+    def reserve_with_range(input_range)
+      @rooms.each do |room|
+        return room.room_id if room.create_room_reservation(input_range)
+      end
+
+      raise ArgumentError.new("No rooms available for date range.")
```
Comment (dHelmgren) [inline lib/hotel_controller.rb:29]: I'd argue that this isn't _really_ an `ArgumentError`. The argument's fine, but your hotel is full. This is a great place for a custom exception!

File: lib/room.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      @rez_list = []
+
+      raise ArgumentError if room_id <= 0 || room_id > 20
+    end
+
+    def add_room_reservation(reservation_to_add)
+      @rez_list << reservation_to_add
+    end
+
+    def conflict?(input_range)
```
Comment (dHelmgren) [inline lib/room.rb:18]: You have mostly got really good method names, but `conflict?` is not quite descriptive enough here. Your asking the room if it conflicts, but that isn't how we'd ask that question conversationally. I'd opt for `is_booked?` or something similar, that communicates the _nature_ of the conflict.

**Comment by dHelmgren** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | ✔️
Practices git with at least 15 small commits and meaningful commit messages | ✔️

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | ✔️
All of the given tests run and pass | ✔️
A test coverage tool is installed and used, and shows 95% test coverage | ✔️

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | ✔️
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | ✔️
calculates the total price for a reservation | ✔️
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | ✔️
creates a block of rooms |
reserves a room from a block |

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections | ✔️
| Yellow (Approaches Standards) | 9-13 total in all sections |
| Red (Not at Standard) | 0-8 total in all sections, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |

### Additional Feedback

Great work overall! You've built your first project with minimal starting code. This represents an incredible milestone in your journey, and you should be proud of yourself!

I am particularly impressed by the way that you organized your code! I found it very easy to see where work was being done in each class, and the way you broke down the problem made it clear that your design efforts were successful! I'm also very keen on how thoroughly you tested the DateRange class!

I do see some room for improvement around naming conventions for methods and errors. Coding, like math, is as much the art of solving problems as it is the art of telling people _how_ you solved them. Names that make it clear what's happening in your methods and to which data can make a huge difference!

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Concise | ✅
| Logical/Organized | ✅
